lib/parseBinaryGltf.js

```diff
@@ -0,0 +1,60 @@
+'use strict';
+var fs = require('fs');
+var Cesium = require('cesium');
+var defined = Cesium.defined;
+var defaultValue = Cesium.defaultValue;
+var StringDecoder = require('string_decoder').StringDecoder;
+module.exports = parseBinaryGltf;
+
+function parseBinaryGltf(data) {
+    var sizeOfUint32 = Uint32Array.BYTES_PER_ELEMENT;
+    var byteOffset = 0;
+
+    if (data.slice(byteOffset, sizeOfUint32).toString() !== 'glTF') {
+        throw new Error('File is not valid binary glTF');
+    }
+
+    byteOffset += 3*sizeOfUint32;
+    var dataView = new DataView(data.buffer);
+    var sceneLength = dataView.getUint32(byteOffset, true);
+
+    byteOffset += 2*sizeOfUint32;
+    var decoder = new StringDecoder();
+    var scene = decoder.write(data.slice(byteOffset, byteOffset + sceneLength));
+    var gltf = JSON.parse(scene);
+
+    byteOffset += sceneLength;
+    var body = data.slice(byteOffset);
+    var buffers = gltf.buffers;
+    if (defined(buffers)) {
+        if (defined(buffers.KHR_binary_glTF || defined(buffers.binary_glTF))) {
+            var buffer = defaultValue(buffers.KHR_binary_glTF, buffers.binary_glTF);
+            buffer.extras = defaultValue(buffer.extras, {});
+            buffer.extras.source = body;
+        }
+    }
+
+    if (defined(gltf.extensionsUsed) && gltf.extensionsUsed.indexOf("KHR_binary_glTF") > -1) {
+        loadSourceFromBody(gltf, body, 'images');
+        loadSourceFromBody(gltf, body, 'shaders');
+    }
+
+    return gltf;
+}
+
+function loadSourceFromBody(gltf, body, name) {
+    var objects = gltf[name];
+    if (defined(objects)) {
+        for (var objectId in objects) {
+            var object = objects[objectId];
+            var objectExtensions = object.extensions;
+            if (defined(objectExtensions) && defined(objectExtensions.KHR_binary_glTF)) {
+                var objectView = objectExtensions.KHR_binary_glTF.bufferView;
+                var bufferView = gltf.bufferViews[objectView];
+                var source = body.slice(bufferView.byteOffset, bufferView.byteOffset + bufferView.byteLength);

```

Comment:
Ah, I see what you were asking in your email now about storing the buffer.

When we write back out the glTF asset, we don't want to duplicate any data in buffers.  We also want to be able to convert non-binary glTF to binary glTF, and perform all optimizations in a generic manner regardless of if the source is binary glTF or not.

Given all this, perhaps the best approach is to not actually store the full `"binary_glTF"` buffer in our in-memory data structure, and instead extract a new buffer for each buffer view (includes those in the extension for images and shaders pointing to buffer views).

Then, we can add a "combine buffers" optimization stage that combines all buffers into one buffer (taking into account if two buffer views point to the same exact part of a buffer, or if a buffer view points to a subset of a buffer pointed to by another buffer view, e.g., we want to avoid duplicating data).

What do you think of this approach?